### PR TITLE
Test cleanup

### PR DESCRIPTION
**Pull request by lluiscab**
I've cleaned up the test files in preparation for the non-vanilla server tests. Test output is now cleaner and it's easier to see what's going on.

### CODE REVIEW

**Comment by lluiscab** [comment]
Seems like the parallel detection is broken or something like that as it's only testing the first two versions. I don't really have an idea of how that file should work, any help @rom1504 ?

**Comment by rom1504** [comment]
No the parallel detection is working as expected. It runs 2 versions by container.
Each container is represented by a tab in circle ci (there are 4 containers)

File: test/benchmark.js [hunk 1]
```diff
@@ -13,39 +13,32 @@ const testDataWrite = [
...
   const version = mcData.version
-  describe('benchmark ' + version.minecraftVersion, function () {
+  describe(`Benchmark for ${version.minecraftVersion}`, function () {
     this.timeout(60 * 1000)
     const inputData = []
-    it('bench serializing', function (done) {
+    it('Bench serializing', function (done) {
       const serializer = mc.createSerializer({state: states.PLAY, isServer: false, version: version.minecraftVersion})
-      let start, i, j
-      console.log('Beginning write test')
```
Comment (rom1504) [inline test/benchmark.js]: Why did you remove these console.log ?

File: test/clientTest.js [hunk 2]
```diff
@@ -25,11 +24,9 @@ mc.supportedVersions.forEach(function (supportedVersion, i) {
     minMem: 1024,
     maxMem: 1024
   })
-  wrap.on('line', function (line) {
-    console.log(line)
-  })
+  // wrap.on('line', console.log);
```
Comment (rom1504) [inline test/clientTest.js]: Why did you remove this ? This is very useful to know what didn't work in a test.

File: test/clientTest.js [hunk 3]
```diff
@@ -25,11 +24,9 @@ mc.supportedVersions.forEach(function (supportedVersion, i) {
...
 
-  describe('client ' + version.minecraftVersion, function () {
+  describe(`Testing minecraft client ${version.minecraftVersion}`, function () {
```
Comment (rom1504) [inline test/clientTest.js]: Should be "Testing minecraft-protocol client" (this is using the minecraft server)

File: test/clientTest.js [hunk 4]
```diff
@@ -41,31 +38,31 @@ mc.supportedVersions.forEach(function (supportedVersion, i) {
       })
     })
 
-    describe('offline', function () {
+    describe('Offline mode tests', function () {
       before(function (done) {
-        console.log(new Date() + 'starting server ' + version.minecraftVersion)
+        console.log(`      Starting minecraft server for ${version.minecraftVersion}`)
```
Comment (rom1504) [inline test/clientTest.js]: Please leave the date, it is useful to see what took time

File: test/clientTest.js [hunk 5]
```diff
@@ -41,31 +38,31 @@ mc.supportedVersions.forEach(function (supportedVersion, i) {
...
         wrap.startServer({
           'online-mode': 'false',
           'server-port': PORT,
-          'motd': 'test1234',
+          'motd': MOTD,
           'max-players': 120
         }, function (err) {
           if (err) { console.log(err) }
-          console.log(new Date() + 'started server ' + version.minecraftVersion)
+          console.log(`      Started minecraft server for ${version.minecraftVersion}`)
```
Comment (rom1504) [inline test/clientTest.js]: Leave the date

File: test/clientTest.js [hunk 6]
```diff
@@ -153,15 +151,14 @@ mc.supportedVersions.forEach(function (supportedVersion, i) {
...
-      it('produce a decent error when connecting with the wrong version', function (done) {
+      it('Produces a decent error when connecting with the wrong version', function (done) {
         const client = mc.createClient({
           username: 'Player',
           version: version.minecraftVersion === '1.8.8' ? '1.11.2' : '1.8.8',
           port: PORT
         })
         client.once('error', function (err) {
           if (err.message.startsWith('This server is version')) {
-            console.log(new Date() + 'Correctly got an error for wrong version : ' + err.message)
```
Comment (rom1504) [inline test/clientTest.js]: Don't remove this

File: test/clientTest.js [hunk 7]
```diff
@@ -237,20 +234,15 @@ mc.supportedVersions.forEach(function (supportedVersion, i) {
...
         const client = mc.createClient({
           username: 'Player',
           version: version.minecraftVersion,
           port: PORT
         })
         client.on('error', err => done(err))
-        let gotKicked = false
         client.on('disconnect', function (packet) {
           assert.ok(packet.reason.indexOf('"Failed to verify username!"') !== -1 || packet.reason.indexOf('multiplayer.disconnect.unverified_username') !== -1)
-          gotKicked = true
```
Comment (rom1504) [inline test/clientTest.js]: Why ?

File: test/common/parallel.js [hunk 8]
```diff
@@ -14,7 +14,7 @@ function testedRange (nodeIndex, nodeTotal, numberOfVersions) {
 
   return {firstVersion: nodeIndex * nbFirsts, lastVersion: (nodeIndex + 1) * nbFirsts - 1}
 }
-console.log({nodeIndex, nodeTotal, versions: mc.supportedVersions.length})
+// console.log({nodeIndex, nodeTotal, versions: mc.supportedVersions.length})
 const {firstVersion, lastVersion} = parallel ? testedRange(nodeIndex, nodeTotal, mc.supportedVersions.length) : {firstVersion: 0, lastVersion: mc.supportedVersions.length - 1}
```
Comment (rom1504) [inline test/common/parallel.js]: Use for debugging

File: test/clientTest.js (hunk 2 above)
Comment (lluiscab) [reply test/clientTest.js]: Tests shouldn't log useless data, they should provide a clean output so the user can see what's going on. It's not really useful to have the minecraft server log. Don't really know why you would want that

File: test/benchmark.js (hunk 1 above)
Comment (lluiscab) [reply test/benchmark.js]: What's the reason to have them?

File: test/clientTest.js (hunk 3 above)
Comment (lluiscab) [reply test/clientTest.js]: Will update with more precise texts

File: test/clientTest.js (hunk 4 above)
Comment (lluiscab) [reply test/clientTest.js]: If you want to know the startup time, then extract it from the server log and print it with the server started text.

File: test/clientTest.js (hunk 5 above)
Comment (lluiscab) [reply test/clientTest.js]: If you want to know the startup time, then extract it from the server log and print it with the server started text.

File: test/clientTest.js (hunk 6 above)
Comment (lluiscab) [reply test/clientTest.js]: Why not? In any case we should log something when we get an unexpected error, we shouldn't log "everything is going ok" messages

File: test/clientTest.js (hunk 7 above)
Comment (lluiscab) [reply test/clientTest.js]: There's no reason to wait for the bot to end, therefore the `gotKicked` variable is not needed.

File: test/common/parallel.js (hunk 8 above)
Comment (lluiscab) [reply test/common/parallel.js]: Could be changed into a formatted console.log message

File: test/benchmark.js (hunk 1 above)
Comment (rom1504) [reply test/benchmark.js]: To know how much time the benchmark take ? It's a benchmark !

File: test/clientTest.js (hunk 2 above)
Comment (rom1504) [reply test/clientTest.js]: This is the most useful data in the client tests. The client tests are testing against the vanilla server so we want to know everything we can.

In practice it was possible to catch several problems using the vanilla server logs (recently that a test was stuck even though the vanilla server correctly saw the player log in)

File: test/clientTest.js (hunk 4 above)
Comment (rom1504) [reply test/clientTest.js]: How is that better ?
This is 100% reliable even if the minecraft wrap or the vanilla server have bugs (which is not that uncommon)

File: test/common/parallel.js (hunk 8 above)
Comment (rom1504) [reply test/common/parallel.js]: Feel free to improve how it looks but the information should be printed

File: test/clientTest.js (hunk 7 above)
Comment (rom1504) [reply test/clientTest.js]: ???
The purpose of the test is to make sure we get kicked and exit properly, so yes this is required.

File: test/clientTest.js (hunk 6 above)
Comment (rom1504) [reply test/clientTest.js]: Not convinced. I only recently fixed this test and this message was very useful to know there was no problem with the test and actually there was a problem with some other unrelated error happening here.

**Comment by rom1504** [comment]
As said we want to have as much info as possible because it's very common these tests fail.

If anything we are not display enough logs. I think we should also display the debug log but circle ci doesn't support that many logs. Could be saved to a file for circle ci to display some better way (would need to look at circle ci doc)

File: test/benchmark.js (hunk 1 above)
Comment (lluiscab) [reply test/benchmark.js]: Mocha already tells you

File: test/clientTest.js (hunk 4 above)
Comment (lluiscab) [reply test/clientTest.js]: Then use start and end times and get the difference. Printing random dates to console makes things hard to read.

File: test/clientTest.js (hunk 7 above)
Comment (lluiscab) [reply test/clientTest.js]: Not if you only finish the test when you get disconnected and the reason is that you got kicked. Otherwise send an error. Make things simple

**Comment by lluiscab** [comment]
Let's say a test fails, it fails becauae you've called the `done` function with an error, you get that error logged onto the console when mocha ends, don't really know why we'd log the minecraft server or some random dates

File: test/clientTest.js (hunk 4 above)
Comment (rom1504) [reply test/clientTest.js]: Well it's not random it's the continuity of all the server log printing the time. But yes we could also print the difference

File: test/clientTest.js (hunk 7 above)
Comment (rom1504) [reply test/clientTest.js]: "disconnect" event isn't "you got disconnected", it's a packet.
We received that packet, mark that we got kicked, and then when the client end make sure we indeed for that packet before getting disconnected

**Comment by rom1504** [comment]
Because the error in the callback is often not enough information to know what is wrong.

**Comment by rom1504** [comment]
Tests aren't only there to say "something doesn't work".
They are also there to say "it fails for that and that reasons, here are some clues how to fix it"

**Comment by rom1504** [comment]
Let's redo or close this.
